### PR TITLE
 Improve the dashboard UX #52 

### PR DESCRIPTION
**Pull request by Farcasut**
I have added a new macro that introduces an additional element to all the dashboard gadgets composed of Analytics macros. The new element is an anchor that redirects the user to the individual page of that macro. When hovered, it will display a description of the macro. Currently, I have included this macro in the 'Analytics category', but I am uncertain whether users should be able to utilize this macro as it only functions when the macros are placed in a dashboard.
Preview:
image

### CODE REVIEW

**Review by mflorea — CHANGES_REQUESTED** [review]
Hi @Farcasut, while I understand the need expressed in #52, I'm a bit surprised by the implementation choice. Can you explain why you chose to use a macro? Did you evaluate other alternative solutions? How is this macro used by the user? Macros are rendering transformations, but they are normally localized, meaning that the transformation is done in the place where the macro is called, so that the macro output is correctly **delimited** and **protected**. Making changes / transformations in other places can cause serious editing issues. For instance, you can end up with those transformations saved in the wiki syntax.

Moreover, it feels wrong for the "description" macro to know about all analytics macros. It should be the other way around. Each analytics macro should be responsible for specifying its description and some shared code is responsible for displaying it. In other words

* either you have a generic macro to display an icon and a message on hover, and each analytics gadget calls this macro in the gadget title
* or each analytics gadget outputs its description hidden somewhere (e.g. as a ``data`` attribute) and you have some common JavaScript / CSS that shows that description after the gadget title

In any case it should be the gadget / macro that decides whether to show or not the icon and what text to display on hover.

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml [hunk 1]
```diff
@@ -131,6 +131,24 @@ analytics.config.checkConnection.error=Failed to connect to Matomo. Please check
...
+analytics.descriptionsAndLinks.configurationsMacro=This report shows the brands / manufacturers of the devices your visitors were using. In most cases this information is only available for non-desktop devices.
+analytics.descriptionsAndLinks.deviceBrandMacro=This report shows the brands / manufacturers of the devices your visitors were using. In most cases this information is only available for non-desktop devices.
+analytics.descriptionsAndLinks.deviceModelMacro=This report shows the devices your visitors are using. Each model is displayed combined with the device brand as some model names are used by multiple brands.
+analytics.descriptionsAndLinks.deviceTypeMacro=This report shows the types of devices your visitors were using. This report will always show all device types Matomo is able to detect, even if there were no visits with a specific type.
+analytics.descriptionsAndLinks.mostViewedMacro=When visitors search on your website, they are looking for a particular page, content, product, or service. This report lists the pages that were clicked the most after an internal search.
+rendering.macro.mostViewedPages.name=Most viewed pages
```
[the remark this answers is not in the document]
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:142]: The translations are added in a separate pull request https://github.com/xwikisas/application-analytics/pull/37.

**Comment by oanalavinia** [comment]
> or each analytics gadget outputs its description hidden somewhere (e.g. as a data attribute) and you have some common JavaScript / CSS that shows that description after the gadget title

For me, option 2 would be best. This allows users to simply use analytics macros on separate dashboards without losing the info icon if they don't add it manually.

Also, make sure to display a title + info icon also when the analytics macros are used outside of a dashboard

**Comment by Farcasut** [comment]
> Hi @Farcasut, while I understand the need expressed in #52, I'm a bit surprised by the implementation choice. Can you explain why you chose to use a macro? Did you evaluate other alternative solutions? How is this macro used by the user? Macros are rendering transformations, but they are normally localized, meaning that the transformation is done in the place where the macro is called, so that the macro output is correctly **delimited** and **protected**. Making changes / transformations in other places can cause serious editing issues. For instance, you can end up with those transformations saved in the wiki syntax.
> 
> Moreover, it feels wrong for the "description" macro to know about all analytics macros. It should be the other way around. Each analytics macro should be responsible for specifying its description and some shared code is responsible for displaying it. In other words
> 
>     * either you have a generic macro to display an icon and a message on hover, and each analytics gadget calls this macro in the gadget title
> 
>     * or each analytics gadget outputs its description hidden somewhere (e.g. as a `data` attribute) and you have some common JavaScript / CSS that shows that description after the gadget title
> 
> 
> In any case it should be the gadget / macro that decides whether to show or not the icon and what text to display on hover.

I took this approach of creating a macro to add the information button to all the analytics macros because I thought it was a simple solution to solve the issue, while also being easy to expand when adding other macros. However, it never occurred to me that this approach could break the view by saving XWiki syntax. As for a better implementation, I'll go with the second option because Lavina also suggested it.

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Description&Links.xml [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+      <contentType>No content</contentType>
+    </property>
+    <property>
+      <defaultCategory>Analytics</defaultCategory>
+    </property>
+    <property>
+      <description/>
+    </property>
+    <property>
+      <id>description</id>
```
[the remark this answers is not in the document]
Comment (trrenty) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/Description&Links.xml:512]: You changed the name of the page but you forgot to change the `id` of the macro.

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Description&Links.xml [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+    </class>
+    <property>
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>require(['jquery'], function ($) {
+
+    function getInfoIcon() {
+        const iconURL = `${XWiki.contextPath}/rest/wikis/${XWiki.currentWiki}/iconThemes/icons?name=info`;
+        return window.fetch(iconURL, {
```
Comment (trrenty) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Description&Links.xml:131]: you chould have used `getJSON` of jQuery https://api.jquery.com/jQuery.getJSON/. See https://github.com/xwiki/xwiki-platform/blob/HEAD/xwiki-platform-core/xwiki-platform-attachment/xwiki-platform-attachment-picker/xwiki-platform-attachment-picker-webjar/src/main/webjar/solrSearch.js#L117-L128 as an example.

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml [hunk 4]
```diff
@@ -44,7 +44,6 @@ analytics.mostViewedPages.header.date=Date
 analytics.mostViewedPages.header.emptyvalue=-
 analytics.mostViewedPages.header.entryNbActions.hint=The number of actions performed by the users.
 analytics.mostViewedPages.header.entryNbActions=No. of actions
-analytics.mostViewedPages.header.exitRate.hint=The percentage of visits that left the website after viewing this page.
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:47]: Where these translations deleted by mistake?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/PagesFollowingASiteSearch.xml [hunk 5]
```diff
@@ -268,7 +269,7 @@
       }
     }
   })
-  #getAnalyticsMacroId('PagesFollowingASiteSearch', $macroId)
+  #getAnalyticsMacroId('pagesFollowingASiteSearchMacro', $macroId)
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/PagesFollowingASiteSearch.xml:272]: Why you added the macro in the id? It does not seem necessary

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 6]
```diff
@@ -190,6 +190,25 @@
   #set ($macroId = $macro + '_' + $analyticsId)
 #end
 
+#macro(displayAnalyticsMacroTitle, $macroReferencer, $descriptionTranslationKey, $nameTranslationKey)
+  #set ($discard = $xwiki.ssx.use('Analytics.Code.Macros.Description&amp;amp;Links'))
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:194]: This page has some JS that has to be loaded on each macro. This seems to be something that might be needed in the future, not only for description and links. So best is to just have a `Analytics.Code.AnalyticsJS` page that is loaded on each macro

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Description&Links.xml [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+      <parse/>
+    </property>
+    <property>
+      <use/>
+    </property>
+  </object>
+  <object>
+    <name>Analytics.Code.Macros.Description&amp;Links</name>
+    <number>0</number>
+    <className>XWiki.WikiMacroClass</className>
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Description&Links.xml:292]: Do you still need the macro?

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 8]
```diff
@@ -190,6 +190,25 @@
...
+  #set ($url = $xwiki.getURL($macroReferencer))
+  #set ($xwiki_gadget = $doc.getObject('XWiki.GadgetClass'))
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:196]: camelCase should be used. WDYT about `isDashboard`?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 9]
```diff
@@ -190,6 +190,25 @@
...
+  #set ($description = $escapetool.xml($services.localization.render($descriptionTranslationKey)))
+  #if ($xwiki_gadget)
+    ## inside a dashboard so we need to add the eye icon to the tile of the dashboard gadget.
+    {{html}}
+      &lt;div class="hidden analytics-description-hidden" data-url=$url data-description="$description"&gt;&lt;/div&gt;
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:201]: This can be a span

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 10]
```diff
@@ -190,6 +190,25 @@
...
+    {{/html}}
+    #set ($discard = $xwiki.jsx.use('Analytics.Code.Macros.Description&amp;amp;Links'))
+  #else
+  ## Needs css, for the time being is just a prototype
+    {{html}}
+     &lt;h2&gt;&lt;div&gt;&lt;span&gt;$escapetool.xml($services.localization.render($nameTranslationKey))&lt;/span&gt; &lt;a title="$description" href=$url&gt;$services.icon.renderHTML('info')&lt;/a&gt;&lt;div&gt;&lt;/h2&gt;
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:207]: Longer than 120ch

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/PagesFollowingASiteSearch.xml (hunk 5 above)
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/PagesFollowingASiteSearch.xml:272]: I added it for consistency with the other analytics macros, and for clarity. However, if you think it's redundant, I can remove it.

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Description&Links.xml (hunk 7 above)
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/Description&Links.xml:292]: No, I forgot to delete it.

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml (hunk 4 above)
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:47]: Yes, the translations were deleted by mistake

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsJS.xml [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+      </use>
+    </class>
+    <property>
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>require(['jquery'], function ($) {
+  function getInfoIcon() {
+    const iconURL = `${XWiki.contextPath}/rest/wikis/${XWiki.currentWiki}/iconThemes/icons?name=info`;
+    return $.ajax({
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsJS.xml:130]: You can use [getJSON](https://api.jquery.com/jquery.getjson/) directly

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/PagesFollowingASiteSearch.xml (hunk 5 above)
Comment (oanalavinia) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/PagesFollowingASiteSearch.xml:272]: Yes, for me it seems redundant

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml [hunk 12]
```diff
@@ -190,7 +190,7 @@ rendering.macro.screenResolution.parameter.startDate.name=End date
 rendering.macro.screenResolution.parameter.startDate.description=Start time of the date interval used for this macro.
 
 ## Admin section
-Analytics.Code.ConfigurationClass_authToken.hint=Token generated by Matomo to be capable of using the API to view the data. This token can be generated from Settings -&gt; Personal -&gt; Security -&gt; Create New Token
+Analytics.Code.ConfigurationClass_authToken.hint=Token generated by Matomo to be capable of using the API to view the data. This token can be generated from Settings -&amp;gt; Personal -&amp;gt; Security -&amp;gt; Create New Token
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:193]: I'm not sure why is this change needed

Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:193]: It is not necessary, but somehow I double-encoded the text

**Comment by Farcasut** [comment]
Previews:
image
image

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsJS.xml [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>require(['jquery'], function ($) {
+  $(document).ready(function () {
+    const iconURL = `${XWiki.contextPath}/rest/wikis/${XWiki.currentWiki}/iconThemes/icons?name=info`;
+    $.getJSON(iconURL, function(response) {
+      const icon = response.icons[0];
+      $('[id^="gadget_"]').each(function () {
+        console.log($(this));
```
Comment (mflorea) [inline application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsJS.xml:133]: Leftover.

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsJS.xml [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+        let hiddenElement = $(this).find('.gadget-content .analytics-description-hidden');
+        if (hiddenElement.length &gt; 0) {
+          let dataElement = $(`&lt;a class="${icon.cssClass} analyticsTitleMargin"&gt;&lt;/a&gt;`);
```
Comment (mflorea) [inline application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsJS.xml:136]: The class name should be semantic, i.e. express what that element is about, e.g. ``analytics-description`` or ``analytics-hint``, etc.

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 15]
```diff
@@ -190,6 +190,25 @@
   #set ($macroId = $macro + '_' + $analyticsId)
 #end
 
+#macro(displayAnalyticsMacroTitle, $macroReferencer, $descriptionTranslationKey, $nameTranslationKey)
```
Comment (mflorea) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:193]:
```suggestion
#macro(displayAnalyticsMacroTitle, $macroDocumentReference, $descriptionTranslationKey, $nameTranslationKey)
```

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 16]
```diff
@@ -190,6 +190,25 @@
...
+  #set ($discard = $xwiki.ssx.use('Analytics.Code.AnalyticsJS'))
+  #set ($url = $xwiki.getURL($macroReferencer))
+  #set ($isDashboard = $doc.getObject('XWiki.GadgetClass'))
+  #set ($description = $escapetool.xml($services.localization.render($descriptionTranslationKey)))
+  #if ($isDashboard)
+    ## inside a dashboard so we need to add the eye icon to the tile of the dashboard gadget.
```
Comment (mflorea) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:199]:
```suggestion
    ## Inside a dashboard so we need to add the eye icon to the tile of the dashboard gadget.
```

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 17]
```diff
@@ -190,6 +190,25 @@
...
+    {{html}}
+      &lt;div class="hidden analytics-description-hidden" data-url=$url data-description="$description"&gt;&lt;/div&gt;
```
Comment (mflorea) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:201]: * the ``hidden`` attribute is preferred over the ``.hidden`` CSS class
* the class name should be semantic; why not use simply ``analytics-description``?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 18]
```diff
@@ -190,6 +190,25 @@
...
 #end
 
+#macro(displayAnalyticsMacroTitle, $macroReferencer, $descriptionTranslationKey, $nameTranslationKey)
+  #set ($discard = $xwiki.ssx.use('Analytics.Code.AnalyticsJS'))
+  #set ($url = $xwiki.getURL($macroReferencer))
+  #set ($isDashboard = $doc.getObject('XWiki.GadgetClass'))
+  #set ($description = $escapetool.xml($services.localization.render($descriptionTranslationKey)))
+  #if ($isDashboard)
+    ## inside a dashboard so we need to add the eye icon to the tile of the dashboard gadget.
+    {{html}}
```
Comment (mflorea) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:200]:
```suggestion
    {{html clean="false"}}
```

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 19]
```diff
@@ -190,6 +190,25 @@
...
+      &lt;div class="hidden analytics-description-hidden" data-url=$url data-description="$description"&gt;&lt;/div&gt;
+    {{/html}}
+    #set ($discard = $xwiki.jsx.use('Analytics.Code.AnalyticsJS'))
+  #else
+  ## Needs css, for the time being is just a prototype
```
Comment (mflorea) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:205]:
```suggestion
    ## Needs css, for the time being is just a prototype
```

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 20]
```diff
@@ -190,6 +190,25 @@
...
+    {{html}}
```
Comment (mflorea) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:206]:
```suggestion
    {{html clean="false"}}
```

**Review by mflorea — CHANGES_REQUESTED** [review]

**Review by mflorea — APPROVED** [review]
Well done!

**Review by oanalavinia — APPROVED** [review]
